server/src/lsp-server.ts

```diff
@@ -674,19 +675,30 @@ export class LspServer {
         }
     }
 
-    async codeAction(params: lsp.CodeActionParams): Promise<(lsp.Command | lsp.CodeAction)[]> {
+    async codeAction(params: lsp.CodeActionParams): Promise<lsp.CodeAction[]> {
         const file = uriToPath(params.textDocument.uri);
         this.logger.log('codeAction', params, file);
         if (!file) {
             return [];
         }
         const args = toFileRangeRequestArgs(file, params.range);
-        const codeActions: (lsp.Command | lsp.CodeAction)[] = [];
-        const errorCodes = params.context.diagnostics.map(diagnostic => Number(diagnostic.code));
-        provideQuickFix(await this.getCodeFixes({ ...args, errorCodes }), codeActions, this.documents);
-        provideRefactors(await this.getRefactors(args), codeActions, args);
-        provideOrganizeImports(file, params.context, codeActions);
-        return codeActions;
+        const actions: lsp.CodeAction[] = [];
+        if (!params.context.only || params.context.only.findIndex((k: string) => k.startsWith(CodeActionKind.QuickFix)) > -1) {
+            const errorCodes = params.context.diagnostics.map(diagnostic => Number(diagnostic.code));
+            actions.push(...provideQuickFix(await this.getCodeFixes({ ...args, errorCodes }), this.documents));
+        }
+        if (!params.context.only || params.context.only.findIndex((k: string) => k.startsWith(CodeActionKind.Refactor)) > -1) {
+            actions.push(...provideRefactors(await this.getRefactors(args), args));
+        }
+
+        // organize import is provided by tsserver for any line, so we only get it if explicitly requested
+        if (params.context.only && params.context.only.findIndex((k: string) => k.startsWith(CodeActionKind.SourceOrganizeImports)) > -1) {
```

Comment:
It seems like the "startsWith" is used "backwards".  For example, the user-provided `only` token with an example value `source.organizeImports.pyright` should not result in this server providing the action since the user-requested one is more specific than `source.organizeImports` (`CodeActionKind.SourceOrganizeImports`).

But also `startsWith` is not a great solution either way because we want to match against full tokens. So `source.fixAll` should match `source.fixAll.eslint` but not `source.fixAllFoo`. So ideally we would split the token at dot and compare the full tokens.

Reply:
Can you clarify a bit?

> `source.organizeImports.pyright` should not result in this server providing the action since the user-requested one is more specific than `source.organizeImports`
> 
> `source.fixAll` should match `source.fixAll.eslint` but not `source.fixAllFoo`

I think you're referring to the different between organize imports and other actions (that we only show organize imports if explicitly requested) but I'm not sure. I'll push up what I'm interpreting this as.

Reply:
Does this comment clarify it https://github.com/theia-ide/typescript-language-server/pull/170#discussion_r661792138 ?

I've touched on two different issues here. The latter is fixed, the former is what I've linked to now.

Reply:
Thanks, that clarifies!